### PR TITLE
Refactor home api

### PR DESCRIPTION
**Pull request by aSvetenco**
in this PR 
1. Presenter and model refactor for achieving testable state;
2. Unit tests for presenters and HomeActivityModel 
3. Here is a beginning of database implementation (because was started before testing) - not finished for now.  Currently, it is 'clean' SQLite - for more deeply understanding and learning purposes. Later it will be converted on ROOM. So if you have some suggests feel free to commit database implementation too.

### CODE REVIEW

File: MeetUp/app/src/main/java/com/endava/meetup/data/db/MeetUpDBHelper.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package com.endava.meetup.data.db;
+
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/data/db/MeetUpDBHelper.java]: i think we discussed implementing Room DB?

File: MeetUp/app/src/main/java/com/endava/meetup/data/db/contract/GroupsContract.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.endava.meetup.data.db.contract;
+
+import android.provider.BaseColumns;
+
+public final class GroupsContract {
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/data/db/contract/GroupsContract.java]: What is Contract?

File: MeetUp/app/src/main/java/com/endava/meetup/screens/auth/modules/intro/IntroFragment.java [hunk 3]
```diff
@@ -67,7 +67,6 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         IntroPresenterFactory introPresenterFactory = new IntroPresenterFactory();
         presenter = introPresenterFactory.getPresenter();
         presenter.attachView(this);
-        presenter.onCheckPermission();
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/screens/auth/modules/intro/IntroFragment.java]: why is this removed?

File: MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivity.java [hunk 4]
```diff
@@ -77,10 +78,13 @@ private HomePresenterFactory getPresenterFactory() {
         User user = getIntent().getParcelableExtra(USER_KEY);
         ConnectivityManager connectivityManager =
                 (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
-        HomeModelFactory modelFactory = new HomeModelFactory(user,
+        MeetUpDBHelper meetUpDBHelper = MeetUpDBHelper.getInstance(this);
+        HomeApiServiceFactory homeApiServiceFactory = new HomeApiServiceFactory(connectivityManager,
                 Api.getInstance().getGroupsApi(),
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivity.java]: is the return for getGroupsApi() used somehow?

File: MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivity.java [hunk 5]
```diff
@@ -77,10 +78,13 @@ private HomePresenterFactory getPresenterFactory() {
...
-                Api.getInstance().getTopicsApi(),
-                connectivityManager, apiErrorHandlerFactory);
+                Api.getInstance().getTopicsApi());
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivity.java]: same here

File: MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivityPresenter.java [hunk 6]
```diff
@@ -1,100 +1,59 @@
...
-        });
+    @Override
+    public void onError(String errorMessage) {
+        showError(errorMessage);
     }
 
-    public void cancelNetworkCalls() {
-        model.cancelCalls();
+    private void showError(String errorMessage) {
+        getView().hideLoading();
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivityPresenter.java]: good job on the callbacks ;)

File: MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivityPresenter.java [hunk 7]
```diff
@@ -1,100 +1,59 @@
...
+        getView().showAlert(errorMessage);
     }
 
-    private void checkIfFurtherCallNeeded() {
-        counter++;
-        if (counter == topics.size()) {
-            updateUi();
+    private void getViewData() {
```
Comment (CebotariM) [inline MeetUp/app/src/main/java/com/endava/meetup/screens/home/HomeActivityPresenter.java]: the presenter should not decide if to fetch the data or update the ui. In a ideal scenario, you should have:
1. model.fetchData()
2. from callback, update the ui with what data comes.
The presenter should not know the source of the data, cached, or downloaded

File: MeetUp/app/src/test/java/com/endava/meetup/screens/auth/modules/intro/IntroFragmentPresenterTest.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Before
+    public void setUp() throws Exception {
+        presenter = new IntroFragmentPresenter();
+    }
+
+    @Test
+    public void attachViewPermissionsGranted() {
+        Mockito.when(view.isPermissionGranted()).thenReturn(true);
+        presenter.attachView(view);
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/auth/modules/intro/IntroFragmentPresenterTest.java]: why don't you attach the view in setUp?

File: MeetUp/app/src/test/java/com/endava/meetup/screens/auth/modules/signin/SignInPresenterTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    public void setUp() throws Exception {
+        presenter = new SignInPresenter(emailValidator,
+                passwordValidator,
+                emptyValidator,
+                model);
+    }
+
+    @Test
+    public void onEmailTextChangedEmptyData() {
+        Mockito.when(emptyValidator.isValid(EMPTY_STRING)).thenReturn(true);
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/auth/modules/signin/SignInPresenterTest.java]: a good practice is not to use empty strings, but actual test strings. Put something related to the app

File: MeetUp/app/src/test/java/com/endava/meetup/screens/auth/modules/signin/SignInPresenterTest.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    @Before
+    public void setUp() throws Exception {
+        presenter = new SignInPresenter(emailValidator,
+                passwordValidator,
+                emptyValidator,
+                model);
+    }
+
+    @Test
+    public void onEmailTextChangedEmptyData() {
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/auth/modules/signin/SignInPresenterTest.java]: The convention says the test method names should represent what the test should verify. The name should be something like "shouldClearErrorValidation" or what exactly this tests verifies

File: MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    HomeApiService apiService;
+
+    @Before
+    public void setUp() throws Exception {
+        model = new HomeActivityModel(user, errorHandler, groupsRepository, apiService);
+    }
+
+    @Test
+    public void getUserTopics() {
+        model.getGroupModels();
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java]: what does this test verifiy?

File: MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    @Test
+    public void fetchData() {
+        model.fetchData();
+        Mockito.verify(apiService).setHomeApiCallback(model);
+        Mockito.verify(apiService).getUserGroupsFromNetwork(user.getApiKey());
+    }
+
+    @Test
+    public void onUserGroupResponseNotEmpty() {
+        List<Group> data = Mockito.mock(List.class);
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java]: please don't mock lists. Create lists of mocks

File: MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        Mockito.when(data.isEmpty()).thenReturn(false);
+        model.onUserGroupResponse(data);
+        Mockito.verify(apiService).getUserTopicsFromNetwork(user.getMemberId(), user.getApiKey());
+    }
+
+    @Test
+    public void onUserTopicsResponseNotEmptyMakeRequest() {
+        List<Topic> data = new ArrayList<>();
+        data.add(new Topic());
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java]: why topic is not mocked?

File: MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Test
+    public void onGroupsByTopicIdResponseNotEmptyUpdateUi() {
+        List<Group> data = Mockito.mock(List.class);
+        Mockito.when(data.isEmpty()).thenReturn(false);
+        ApiResultCallback<List<SingleGroupModel>> callback = Mockito.mock(ApiResultCallback.class);
+        model.setApiResultCallback(callback);
+        List<Topic> topics = new ArrayList<>();
+        topics.add(new Topic());
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java]: same here

File: MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    @Test
+    public void onGroupsByTopicIdResponseNotEmptyMakeRequest() {
+        List<Group> data = Mockito.mock(List.class);
+        Mockito.when(data.isEmpty()).thenReturn(false);
+        List<Topic> topics = new ArrayList<>();
+        topics.add(new Topic());
+        topics.add(new Topic());
+        model.setUserTopics(topics);
+        model.onGroupsByTopicIdResponse(data);
+        Mockito.verify(apiService).getGroupsByTopicId(null, "28.8", "47.0", null);
```
Comment (CebotariM) [inline MeetUp/app/src/test/java/com/endava/meetup/screens/home/HomeActivityModelTest.java]: why null is passed here?

File: MeetUp/app/src/main/java/com/endava/meetup/data/db/contract/GroupsContract.java (hunk 2 above)
Comment (aSvetenco) [reply MeetUp/app/src/main/java/com/endava/meetup/data/db/contract/GroupsContract.java]: google in tutorial calls such db schema
